content/docs/rfcs/12/README.md

```diff
@@ -61,77 +63,154 @@ The following are not considered as part of the adversarial model:
+In addition, the filter service node MAY choose to provide a more detailed status description in the `status_desc` field.
+
+### Filter matching
+
+In the description of each request type below,
+the term "filter criteria" refers to the combination of `pubsub_topic` and a set of `content_topics`.
+The request MAY include filter criteria, conditional to the selected `filter_subscribe_type`.
+The `pubsub_topic` is always optional in filter criteria.
+A `WakuMessage` matches filter criteria when its `content_topic` is in the `content_topics` set
+and, if included in the criteria, it was published on a matching `pubsub_topic`.
+If the `pubsub_topic` is set and not empty,
+a `WakuMessage` will only match the filter criteria if it has a matching `content_topic` and was published on _the same_ `pubsub_topic`.
+If the `pubsub_topic` is either empty or unset,
+a `WakuMessage` will match the filter criteria if it has a matching `content_topic` and was published on _any_ `pubsub_topic`.
+
+### Filter Subscribe Types
+
+The following filter subscribe types are defined:
+
+#### SUBSCRIBER_PING
+
+A filter client that sends a `FilterSubscribeRequest` with `filter_subscribe_type` set to `SUBSCRIBER_PING`
+requests that the service node SHOULD indicate if it has any active subscriptions for this client.
+The filter client SHOULD exclude any filter criteria from the request.
+The filter service node SHOULD respond with a success code if it has any active subscriptions for this client
+or an error code if not.
+The filter service node SHOULD ignore any filter criteria in the request.
+
+#### SUBSCRIBE
+
+A filter client that sends a `FilterSubscribeRequest` with `filter_subscribe_type` set to `SUBSCRIBE`
+requests that the service node SHOULD push messages matching this filter to the client.
+The filter client MUST include the desired filter criteria in the request.
+A client MAY use this request type to _modify_ an existing subscription
+by providing _additional_ filter criteria in a new request.
+A client MAY use this request type to _refresh_ an existing subscription
+by providing _the same_ filter criteria in a new request.
+The filter service node SHOULD respond with a success code if it successfully honored this request
+or an error code if not.
+
+#### UNSUBSCRIBE
+
+A filter client that sends a `FilterSubscribeRequest` with `filter_subscribe_type` set to `UNSUBSCRIBE`
+requests that the service node SHOULD _stop_ pushing messages matching this filter to the client.
+The filter client MUST include the filter criteria it desires to unsubscribe from in the request.
+A client MAY use this request type to _modify_ an existing subscription
+by providing _a subset of_ the original filter criteria to unsubscribe from in a new request.
+The filter service node SHOULD respond with a success code if it successfully honored this request
+or an error code if not.
```

Comment:
What happens if I a client sends an `UNSUBSCRIBE` request containing a `pubsub_topic` and no `content_topics`? (since a `repeated` field can contain 0 items)

Reply:
Ah, thanks for picking this up! It should definitely be clarified in the specs. My view is that `SUBSCRIBE` and `UNSUBSCRIBE` (in fact the idea of subscriptions) should be related to content topics, qualified by a (perhaps now mandatory) pubsub topic. That means that only content topics will be considered in subscribe and unsubscribe requests. Not having any content topics will be seen as an empty unsubscribe request and SHOULD be ignored.

Of course, LMK if you believe from client perspective if a different semantic would work better, but the main idea of Filter is to operate on content topics.